src/wh_server_crypto.c

```diff
@@ -0,0 +1,337 @@
+        ret = BUFFER_E;
+    /* read key */
+    if (ret == 0) {
+        ret = hal_flash_read(addr + WOLFHSM_HEADER_SIZE + keyOffset, out,
+            ctx->nvmMetaCache[i].len);
+    }
+    /* calculate key digest */
+    if (ret == 0)
+        ret = wc_InitSha256_ex(sha, ctx->heap, ctx->devId);
+    if (ret == 0)
+        ret = wc_Sha256Update(sha, out, ctx->nvmMetaCache[i].len);
+    if (ret == 0)
+        ret = wc_Sha256Final(sha, digest);
+    /* validate key */
+    if (XMEMCMP(digest, ctx->nvmMetaCache[i].confDigest, WOLFHSM_DIGEST_STUB)
+        != 0) {
+        ret = BAD_FUNC_ARG;
+    }
+    if (ret == 0) {
+        /* cache key if free slot, will only kick out other commited keys */
+        hsmCacheKey(ctx, &ctx->nvmMetaCache[i], out);
+        /* update out meta with nvm cache contents */
+        XMEMCPY(meta, &ctx->nvmMetaCache[i], sizeof(NvmMetaData));
+    }
+#ifdef WOLFHSM_SHE_EXTENSION
+    /* use empty string if we couldn't find the master ecu key */
+    if (ret != 0 && searchId == WOLFHSM_SHE_MASTER_ECU_KEY_ID) {
+        XMEMSET(out, 0, WOLFHSM_SHE_KEY_SZ);
+        meta->len = WOLFHSM_SHE_KEY_SZ;
+        meta->id = searchId;
+        ret = 0;
+    }
+#endif
+    return ret;
+#endif
+}
+
+static int hsmCacheKeyCurve25519(whServerContext* server, curve25519_key* key)
+{
+    int ret;
+    uint32_t privSz = CURVE25519_KEYSIZE;
+    uint32_t pubSz = CURVE25519_KEYSIZE;
+    whNvmMetadata meta[1] = {0};
+    byte keyBuf[CURVE25519_KEYSIZE * 2];
+    /* store public, then private so that loading an external public only key
+     * will work along with our keys */
+    ret = wc_curve25519_export_key_raw(key, keyBuf + CURVE25519_KEYSIZE,
+        &privSz, keyBuf, &pubSz);
+    /* cache key */
+    if (ret == 0) {
+        ret = hsmGetUniqueId(server);
+    }
+    if (ret > 0) {
+        XMEMSET(meta, 0, sizeof(whNvmMetadata));
+        meta->len = privSz + pubSz;
+        meta->id = ret;
+        ret = hsmCacheKey(server, meta, keyBuf);
+    }
+    if (ret == 0)
+        ret = meta->id;
+    return ret;
+}
+
+static int hsmLoadKeyCurve25519(whServerContext* server, curve25519_key* key, uint16_t keyId)
+{
+    int ret;
+    uint32_t privSz = CURVE25519_KEYSIZE;
+    uint32_t pubSz = CURVE25519_KEYSIZE;
+    whNvmMetadata meta[1] = {0};
+    byte keyBuf[CURVE25519_KEYSIZE * 2];
+    /* retrieve the key */
+    XMEMSET(meta, 0, sizeof(whNvmMetadata));
+    meta->id = keyId;
+    meta->len = privSz + pubSz;
+    ret = hsmReadKey(server, meta, keyBuf);
+    /* decode the key */
+    if (ret == 0)
+        ret = wc_curve25519_import_public(keyBuf, pubSz, key);
+    /* only import private if what we got back holds 2 keys */
+    if (ret == 0 && meta->len == CURVE25519_KEYSIZE * 2)
+        ret = wc_curve25519_import_private(keyBuf + pubSz, privSz, key);
+    return ret;
+}
+
+int _wh_Server_HandleCryptoRequest(whServerContext* server,
+    uint16_t* action, uint8_t* data, uint16_t* size)
```

Comment:
Action doesn't tell you what error occurred, it tells the cryptocb layer that the action it wanted wasn't finished and then packet->error gives the specific error code

Reply:
The intent of the Group/Action was to identify the kind of request/response, with the expectation that a request of a specific group+action will result in a single response of that same group+action, even under error cases.  There's a check in comm_client that ensures that the response group+action matches the last request group+action.  The action should NOT be changed on the response under normal and error circumstances.  I haven't completed the async cancel handling yet that would allow an additional request or an unexpected response to be sent and handled correctly.  These would use fixed group+action values to make processing simple.